### PR TITLE
feat(memh5): zarr support

### PR DESCRIPTION
**Pull request by nritsche**
TODO:
- [x] _distributed_group_to_zarr
- [x] unit tests
- [ ] test: format some old hdf5 files in zarr
- [x] run daily processing pipeline with zarr
- [x] we are creating `<filename>.zarr.sync` folders for zarr's synchronizer, but they don't get deleted by zarr. do we have to do that? https://github.com/harlowja/fasteners/issues/26
- [x] optimize striping for small (chunk) files: https://scicomp.aalto.fi/triton/usage/lustre/
- [x] increase chunk size to ~16MB (compressed on disk)

Also see https://github.com/radiocosmology/draco/pull/130
and chime-experiment/ch_pipeline#82

Closes https://github.com/chime-experiment/Pipeline/issues/33

**Edit by @anjakefala**
As of 2021-08-23, here is the discussed remaining work for this PR, along with any unchecked boxes above:
- [x] remove references to `chunkify()`
- [x] Ignore compression and chunking options when saving to `.h5` (log `DEBUG` level)
- [ ] add `zarr` to the base chime environment on Cedar, before we set the pipeline running with this new code
- [x] make sure the docs can build and upload to readthedocs; one way is to either make `bitshuffle` an optional dependency for building the docs, or build the docs and upload them through readthedocs through github actions (see ch_util)
      - Update: ch_util actually pushes the docs to github pages, not readthedocs: https://github.com/chime-experiment/ch_util/pull/10
- [x] clean up commits made by @anjakefala 
- [ ] @jrs65 and @anjakefala review the full PR
- [x] get a successful run with `test_daily` and then get a successful run with the full pipeline.

### CODE REVIEW

File: caput/mpiarray.py [hunk 1]
```diff
@@ -830,7 +882,7 @@ def to_file(
                 islice, fslice = _partition_sel(
                     sel, split_axis, self.global_shape[split_axis], part
                 )
-                dset[islice] = self[fslice]
+                dset[islice] = self.local_array[fslice]
```
Comment (anjakefala) [inline caput/mpiarray.py:885]: yay

**Comment by nritsche** [comment]
https://github.com/chime-experiment/ch_pipeline/pull/82

**Comment by anjakefala** [comment]
By default, `mkchimeenv.sh` currently does not install `zarr` when it installs `caput`. We should perhaps add `zarr` to our chime environment as part of shipping this.

File: caput/memh5.py [hunk 2]
```diff
@@ -411,7 +425,7 @@ def from_hdf5(
         selections=None,
         convert_dataset_strings=False,
         convert_attribute_strings=True,
-        **kwargs
+        **_,
```
Comment (anjakefala) [inline caput/memh5.py:428]: Rick is not passing on the `kwargs` here, I am not sure why, yet. Open question if this that we want to do?

Comment (anjakefala) [reply caput/memh5.py:428]: The more I look at this, the more I think that we do want to pass on the `kwargs`, and removing them was possibly a linting error.

File: caput/memh5.py [hunk 3]
```diff
@@ -484,7 +556,9 @@ def to_hdf5(
         hints=True,
         convert_attribute_strings=True,
         convert_dataset_strings=False,
-        **kwargs
+        compression=None,
+        compression_opts=None,
+        **kwargs,
```
Comment (anjakefala) [inline caput/memh5.py:561]: We may want to add `chunks` as an explicit argument.

File: caput/memh5.py [hunk 4]
```diff
@@ -503,16 +577,61 @@ def to_hdf5(
         convert_dataset_strings : bool, optional
             Try and convert dataset string types to bytestrings. Default is `False`.
```
Comment (anjakefala) [inline caput/memh5.py:578]: And add the `compression` options here in the docstrings.

File: caput/memh5.py (hunk 3 above)
Comment (anjakefala) [reply caput/memh5.py:561]: Oh, chunking does not happen with `MemGroup`?

File: caput/memh5.py [hunk 5]
```diff
@@ -1282,9 +1450,11 @@ class MemDiskGroup(_BaseGroup):
     detect_subclass: boolean, optional
         If *data_group* is specified, whether to inspect for a
         '__memh5_subclass' attribute which specifies a subclass to return.
+    file_format : `fileformats.FileFormat`
+        File format to use. File format will be guessed if not supplied. Default `None`.
     """
 
-    def __init__(self, data_group=None, distributed=False, comm=None):
+    def __init__(self, data_group=None, distributed=False, comm=None, file_format=None):
```
Comment (anjakefala) [inline caput/memh5.py:1457]: I think `file_format` can have a fallback value of `None` in a couple of places, and I am wondering if this ever breaks anything. Testing a form of `.yaml` where you have a `.zarr` filename, but do not specify the `output_format`.

File: caput/memh5.py [hunk 6]
```diff
@@ -2066,28 +2295,28 @@ def get_h5py_File(f, **kwargs):
...
+        if file_format == fileformats.Zarr and not zarr_available:
+            raise RuntimeError("Unable to open zarr file. Please install zarr.")
         try:
-            f = h5py.File(f, **kwargs)
+            f = file_format.open(f, **kwargs)
         except IOError as e:
             msg = "Opening file %s caused an error: " % str(f)
             raise IOError(msg + str(e)) from e
-    return f, opened
+        return f, file_format == fileformats.HDF5
```
Comment (anjakefala) [inline caput/memh5.py:2309]: Why does it not return `True` for `zarr` as well? Are `zarr` files by default "already opened"?

File: caput/memh5.py [hunk 7]
```diff
@@ -2226,27 +2470,40 @@ def deep_group_copy(
             if convert_dataset_strings:
                 # Convert unicode strings back into ascii byte strings. This will break
                 # if there are characters outside of the ascii range
-                if isinstance(g2, h5py.Group):
+                if isinstance(g2, file_format.module.Group):
```
Comment (anjakefala) [inline caput/memh5.py:2473]: This will `Exception` if `not zarr_available`

File: caput/memh5.py [hunk 8]
```diff
@@ -2226,27 +2470,40 @@ def deep_group_copy(
...
+                if compression_entry is not None:
+                    compression = compression_entry
+                    compression_opts = getattr(entry, "compression_opts", None)
+                compression_kwargs = file_format.compression_kwargs(
+                    compression=compression,
+                    compression_opts=compression_opts,
+                    compressor=getattr(entry, "compressor", None),
+                )
+            else:
+                # use HDF5 compression args format
```
Comment (anjakefala) [inline caput/memh5.py:2496]: If this codeblock is going to run for `file_format.Zarr` as well, then we cannot have `HDF5.compression_kwargs` return `{}` by default. I actually do not understand what the `if isinstance(g2, file_format.module.Group)` check is about. Is this, if it is a `dataset`, then we use `HDF5.compression_kwargs`? Does `zarr` have datasets?

File: caput/memh5.py [hunk 9]
```diff
@@ -2471,9 +2727,12 @@ def _copy_to_file(memgroup, h5group):
...
                         dtype=data.dtype,
-                        chunks=item.chunks,
-                        compression=item.compression,
-                        compression_opts=item.compression_opts,
+                        chunks=item.chunks
+                        if fileformats.HDF5.compression_enabled()
+                        else None,
+                        **fileformats.HDF5.compression_kwargs(
+                            item.compression, item.compression_opts
+                        ),
```
Comment (anjakefala) [inline caput/memh5.py:2735]: I wonder why `compression_kwargs` are used here, but not at 2709..

File: caput/memh5.py [hunk 10]
```diff
@@ -2503,13 +2762,135 @@ def _copy_to_file(memgroup, h5group):
...
+                        data = check_unicode(item)
+
+                    # Write common data from rank 0
+                    if memgroup.comm.rank == 0:
+                        dset = group.create_dataset(
+                            key,
+                            shape=data.shape,
+                            dtype=data.dtype,
+                            chunks=item.chunks,
+                            **fileformats.Zarr.compression_kwargs(
```
Comment (anjakefala) [inline caput/memh5.py:2846]: I would say, we should ensure every `create_dataset` has the `fileformats.compression_kwargs` treatment...

File: caput/memh5.py [hunk 11]
```diff
@@ -2503,13 +2762,135 @@ def _copy_to_file(memgroup, h5group):
...
+            f.attrs["__memh5_distributed_file"] = True
+
+    # Final synchronisation
+    group.comm.Barrier()
+
+
+def _distributed_group_from_file(
+    fname,
+    comm=None,
+    _=True,
```
Comment (anjakefala) [inline caput/memh5.py:2889]: `_=True` used to be `hints=True`. I am guessing `hints` is deprecated?

File: caput/memh5.py [hunk 12]
```diff
@@ -2503,13 +2762,135 @@ def _copy_to_file(memgroup, h5group):
...
+    convert_dataset_strings=False,
+    convert_attribute_strings=True,
+    file_format=fileformats.HDF5,
+    **kwargs,
 ):
     """
     Restore full tree from an HDF5 file into a distributed memh5 object.
```
Comment (anjakefala) [inline caput/memh5.py:2896]: "from an HDF5 file or `Zarr group`"

File: caput/memh5.py [hunk 13]
```diff
@@ -2594,10 +2980,14 @@ def _copy_from_file(h5group, memgroup, selections=None):
...
-    with misc.open_h5py_mpi(fname, "r", comm=comm) as f:
+    if file_format == fileformats.HDF5:
+        # Open file on all ranks
+        with misc.open_h5py_mpi(fname, "r", comm=comm) as f:
 
-        # Start recursive file read
+            # Start recursive file read
+            _copy_from_file(f, group, selections)
+    else:
+        f = file_format.open(fname, "r")
```
Comment (anjakefala) [inline caput/memh5.py:2990]: From this I am assuming that this is intended to work with `Zarr` as well

File: caput/mpiarray.py [hunk 14]
```diff
@@ -631,12 +631,61 @@ def from_hdf5(cls, f, dataset, comm=None, axis=0, sel=None):
...
-        use_mpi = axis > 0
+        return cls.from_file(f, dataset, comm, axis, sel, file_format=fileformats.HDF5)
 
-        # Read the file. Opening with MPI if requested, and we can
-        fh = misc.open_h5py_mpi(f, "r", use_mpi=use_mpi, comm=comm)
+    @classmethod
+    def from_file(
+        cls, f, dataset, comm=None, axis=0, sel=None, file_format=fileformats.HDF5
+    ):
+        """Read MPIArray from an HDF5 dataset in parallel.
```
Comment (anjakefala) [inline caput/mpiarray.py:640]: "/Zarr Dataset"

File: caput/mpiarray.py [hunk 15]
```diff
@@ -631,12 +631,61 @@ def from_hdf5(cls, f, dataset, comm=None, axis=0, sel=None):
...
+            # Blosc may share incorrect global state amongst processes causing programs to hang.
+            # See https://zarr.readthedocs.io/en/stable/tutorial.html#parallel-computing-and-synchronization
+            try:
+                import numcodecs
+            except ImportError:
+                raise RuntimeError("Install numcodecs to read from zarr files.")
+            numcodecs.blosc.use_threads = False
+
+            if isinstance(f, str):
+                fh = file_format.open(f, "r")
```
Comment (anjakefala) [inline caput/mpiarray.py:682]: Okay, so we do not open `zarr` with MPI

File: caput/mpiarray.py [hunk 16]
```diff
@@ -712,12 +770,11 @@ def to_hdf5(
 
         Parameters
         ----------
-        filename : str, h5py.File or h5py.Group
+        f : str, h5py.File or h5py.Group
             File to write dataset into.
         dataset : string
```
Comment (anjakefala) [inline caput/mpiarray.py:775]: Add to docstring

File: caput/mpiarray.py [hunk 17]
```diff
@@ -730,39 +787,38 @@ def to_hdf5(
...
         # Only use collective IO if:
         # - there are no null slices (h5py bug)
         # - we are not distributed over axis=0 as there is no advantage for
         #   collective IO which is usually slow
         # - unless we want to use compression/chunking
         # TODO: change if h5py bug fixed
+        # https://github.com/h5py/h5py/issues/965
         # TODO: better would be a test on contiguous IO size
         use_collective = (
             fh.is_mpi and no_null_slices and (self.axis > 0 or compression is not None)
```
Comment (anjakefala) [inline caput/mpiarray.py:805]: On second thought, we may want to include here that compression is turned off for `HDF5`. Because since we are not doing compression, there is a possibility of using collective IO.

File: caput/mpiarray.py [hunk 18]
```diff
@@ -730,39 +787,38 @@ def to_hdf5(
...
         )
 
-        if fh.is_mpi and not use_collective:
+        if (
+            fh.is_mpi
+            and (not use_collective)
+            or (not fileformats.HDF5.compression_enabled())
+        ):
             # Need to disable compression if we can't use collective IO
```
Comment (anjakefala) [inline caput/mpiarray.py:813]: Though, I do not really understand collective IO, and its relationship to compression.

File: caput/mpiarray.py [hunk 19]
```diff
@@ -730,39 +787,38 @@ def to_hdf5(
...
         # TODO: better would be a test on contiguous IO size
         use_collective = (
             fh.is_mpi and no_null_slices and (self.axis > 0 or compression is not None)
         )
 
-        if fh.is_mpi and not use_collective:
+        if (
+            fh.is_mpi
+            and (not use_collective)
+            or (not fileformats.HDF5.compression_enabled())
```
Comment (anjakefala) [inline caput/mpiarray.py:811]: This is where I added the compression enabled check, to turn off `chunks`, and `compression`.

File: caput/mpiarray.py [hunk 20]
```diff
@@ -730,39 +787,38 @@ def to_hdf5(
...
             # Need to disable compression if we can't use collective IO
             chunks, compression, compression_opts = None, None, None
 
-        dset = fh.create_dataset(
+        sel = self._make_selections()
+
+        # Split the axis to get the IO size under ~2GB (only if MPI-IO)
+        split_axis, partitions = self._partition_io(skip=(not fh.is_mpi))
+
+        fh.create_dataset(
```
Comment (anjakefala) [inline caput/mpiarray.py:821]: If `create_dataset` is where Rick has been adding `fileformats.compression_kwargs`, then maybe the better place to turn off chunking et al is within `create_dataset` calls.

File: caput/mpiarray.py [hunk 21]
```diff
@@ -772,18 +828,149 @@ def to_hdf5(
...
+        # Split the axis
+        split_axis, partitions = self._partition_io(skip=True)
+
+        if self.comm.rank == 0:
+            group.create_dataset(
+                dataset,
+                shape=self.global_shape,
+                dtype=self.dtype,
+                chunks=chunks,
+                **extra_args,
```
Comment (anjakefala) [inline caput/mpiarray.py:913]: These are `**Zarr.compression_args()`

File: caput/pipeline.py [hunk 22]
```diff
@@ -1263,7 +1266,13 @@ def read_process_write(self, input, input_filename, output_filename):
             output_dirname = os.path.dirname(output_filename)
             if not os.path.isdir(output_dirname):
                 os.makedirs(output_dirname)
-            self.write_output(output_filename, output)
+            self.write_output(
```
Comment (anjakefala) [inline caput/pipeline.py:1269]: `write_output()` lost its `self` parameter (It is now a `staticmethod`, so I am not quite sure how this has been working for my tests!

File: caput/pipeline.py [hunk 23]
```diff
@@ -1436,34 +1449,51 @@ class H5IOMixin:
     # TODO, implement reading on disk (i.e. no copy to memory).
     # ondisk = config.Property(default=False, proptype=bool)
 
-    def read_input(self, filename):
+    @staticmethod
+    def read_input(filename):
```
Comment (anjakefala) [inline caput/pipeline.py:1453]: I wonder why Rick made all of these `staticmethod`.

File: caput/tests/conftest.py [hunk 24]
```diff
@@ -89,10 +90,51 @@ def read_input(self, filename):
     def read_output(self, filename):
         raise NotImplementedError()
 
-    def write_output(self, filename, output):
+    def write_output(self, filename, output, file_format=None, **kwargs):
```
Comment (anjakefala) [inline caput/tests/conftest.py:93]: Do these need to be `staticmethod`?

File: caput/tod.py [hunk 25]
```diff
@@ -118,7 +120,7 @@ def __init__(self, files):
 
         # Fetch all meta data.
         time = np.copy(data_empty.time)
-        first_file, toclose = memh5.get_h5py_File(files[0])
```
Comment (anjakefala) [inline caput/tod.py:121]: We went from a default of `HDF5` to a default of "guessing from filename". This will be fine if we use a consistent suffix with our hdf5 files.

File: setup.py [hunk 26]
```diff
@@ -61,7 +61,15 @@
     """,
     python_requires=">=3.6",
     install_requires=requires,
-    extras_require={"mpi": ["mpi4py>=1.3"], "profiling": ["psutil", "pyinstrument"]},
+    extras_require={
+        "mpi": ["mpi4py>=1.3"],
+        "compression": [
+            "bitshuffle @ git+https://github.com/kiyo-masui/bitshuffle.git",
+            "numcodecs==0.7.3",
```
Comment (anjakefala) [inline setup.py:68]: The `compression` extra is new, `bitshuffle` and `numcodecs` will not install by default.

File: caput/memh5.py [hunk 27]
```diff
@@ -2276,7 +2533,7 @@ def _distributed_group_to_hdf5_serial(
     hints=True,
     convert_dataset_strings=False,
     convert_attribute_strings=True,
-    **kwargs
+    **kwargs,
```
Comment (anjakefala) [inline caput/memh5.py:2536]: If I understand correctly, the `create_dataset` in `_distributed_group_to_hdf5_serial` is not going through `compression_kwargs` because the bug only pops up for parallel.

Comment (anjakefala) [reply caput/memh5.py:2536]: The compression and truncation bug for HDF5.

File: caput/memh5.py (hunk 5 above)
Comment (anjakefala) [reply caput/memh5.py:1457]: You can remove the `output_format` and `.zarr` outputs are still accurately typed. Guessing this is fine, then.

File: caput/memh5.py (hunk 8 above)
Comment (anjakefala) [reply caput/memh5.py:2496]: I removed the return `{}` by default in `HDF5.compression_kwargs` until we understand this better.

File: caput/memh5.py (hunk 27 above)
Comment (anjakefala) [reply caput/memh5.py:2536]: It was confirmed by Richard that the bug only matters for parallelised IO, but we should still be formatting the `compression_kwargs`, I think, so I will add that.

File: caput/memh5.py (hunk 6 above)
Comment (anjakefala) [reply caput/memh5.py:2317]: `Zarr` fileformats do not necessarily have a `close()` method; we should take this outside of `get_file`, and decide on how that file format gets closed.

File: caput/memh5.py (hunk 8 above)
Comment (anjakefala) [reply caput/memh5.py:2510]: If `else`, that is the `in-memory` case, and by default Rick uses the HDF5 compression kwarg format for that case.

File: caput/memh5.py (hunk 11 above)
Comment (anjakefala) [reply caput/memh5.py:2914]: Leave a comment that it does not do anything in this method.

File: caput/mpiarray.py (hunk 18 above)
Comment (anjakefala) [reply caput/mpiarray.py:814]: We should log that `compression` has been turned off, or throw an Exception.

File: caput/pipeline.py (hunk 23 above)
Comment (anjakefala) [reply caput/pipeline.py:1453]: Probably linter

File: caput/pipeline.py [hunk 28]
```diff
@@ -1263,7 +1266,13 @@ def read_process_write(self, input, input_filename, output_filename):
             output_dirname = os.path.dirname(output_filename)
             if not os.path.isdir(output_dirname):
                 os.makedirs(output_dirname)
-            self.write_output(output_filename, output)
+            _OneAndOne.write_output(
```
Comment (anjakefala) [inline caput/pipeline.py:1269]: Revert this back to `self.write_output`, if `self.` will work with a `staticmethod`.

or `self.__class__.write_output`

File: caput/tod.py (hunk 25 above)
Comment (anjakefala) [reply caput/tod.py:121]: If a guess is not conclusive, maybe hdf5?

File: setup.py [hunk 29]
```diff
@@ -61,7 +61,15 @@
...
+        ],
+        "zarr": ["zarr==2.8.1"],
```
Comment (anjakefala) [inline setup.py:70]: It does not necessarily make sense to separate `zarr` from `numcodecs`, since they are used together.

See if you can add platform specifiers, (Check for MACOS, and if MACOS, do not install `zarr`.

File: caput/pipeline.py (hunk 28 above)
Comment (anjakefala) [reply caput/pipeline.py:1269]: It does!

File: caput/memh5.py [hunk 30]
```diff
@@ -2168,6 +2414,9 @@ def deep_group_copy(
     selections=None,
     convert_dataset_strings=False,
     convert_attribute_strings=True,
+    file_format=fileformats.HDF5,
+    compression=None,
+    compression_opts=None,
```
Comment (jrs65) [inline caput/memh5.py:2419]: Should these be able to be specified per dataset?

File: caput/memh5.py [hunk 31]
```diff
@@ -2330,21 +2602,34 @@ def _distributed_group_to_hdf5_serial(
...
+            else:
+                # disable compression if not enabled for HDF5 files
+                # https://github.com/chime-experiment/Pipeline/issues/33
+                arr.to_hdf5(
+                    fname,
+                    entry.name,
+                    chunks=None,
+                    compression=None,
+                    compression_opts=None,
+                )
```
Comment (jrs65) [inline caput/memh5.py:2632]: It seems like this would be better structured with the conditional formatting the arguments and then a single call to `.to_hdf5`

Comment (jrs65) [reply caput/memh5.py:2632]: Or move the conditional into `.to_hdf5`

File: caput/memh5.py [hunk 32]
```diff
@@ -2445,13 +2733,24 @@ def _copy_to_file(memgroup, h5group):
...
 
                     # Write to file from MPIArray
-                    data.to_hdf5(
-                        h5group,
-                        key,
-                        chunks=item.chunks,
-                        compression=item.compression,
-                        compression_opts=item.compression_opts,
-                    )
+                    if fileformats.HDF5.compression_enabled():
```
Comment (jrs65) [inline caput/memh5.py:2736]: Same comments as above.

Comment (jrs65) [reply caput/memh5.py:2736]: Also, how did we avoid this before? I guess we just relied on the user not setting any compression as it would break.

File: caput/mpiarray.py [hunk 33]
```diff
@@ -712,12 +770,11 @@ def to_hdf5(
 
         Parameters
         ----------
-        filename : str, h5py.File or h5py.Group
+        f : str, h5py.File or h5py.Group
             File to write dataset into.
         dataset : string
             Name of dataset to write into. Should not exist.
```
Comment (jrs65) [inline caput/mpiarray.py:776]: Needs documentation for all parameters.

File: caput/pipeline.py [hunk 34]
```diff
@@ -1330,6 +1340,9 @@ class SingleBase(_OneAndOne):
 
     input_filename = config.Property(default="", proptype=str)
     output_filename = config.Property(default="", proptype=str)
+    output_format = config.file_format()
+    output_compression = config.Property(default=None, proptype=str)
+    output_compression_opts = config.Property(default=None)
```
Comment (jrs65) [inline caput/pipeline.py:1345]: I think we need to review if these settings should be *per* dataset.

File: caput/tests/test_selection_parallel.py [hunk 35]
```diff
@@ -2,59 +2,86 @@
...
 
     comm.Barrier()
 
     # tear down
-
     if comm.rank == 0:
-        file_names = glob.glob(fname + "*")
-        for fname in file_names:
-            os.remove(fname)
+        rm_all_files(zarr_file)
```
Comment (jrs65) [inline caput/tests/test_selection_parallel.py:59]: It does seem like this could factor out the memh5 container creation

**Review by jrs65 — CHANGES_REQUESTED** [review]

**Comment by jrs65** [comment]
I added a bunch of comments, there not all that actionable, some are just notes to myself wondering how something is meant to work.

File: caput/tests/test_truncate.py [hunk 36]
```diff
@@ -0,0 +1,77 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (tristpinsm) [reply caput/tests/test_truncate.py:1]: yeah I guess it could be useful to test the behaviour of some edge cases, if only to be aware of their outcomes. These could include:
* precision set to 0/inf/NaN
* weight set to 0/inf/NaN
* value set to inf/NaN
* combinations of the above
* not an edge case exactly but truncating negative numbers

More generally a more robust test that I was using when developing these functions was to generate an ensemble of random numbers, truncate them, and check that the distribution of truncation errors was what you expect. That might be cumbersome and difficult to implement in this context though.

Another thing to note is there is a comment that claims one of the tests fails...

File: setup.py (hunk 29 above)
Comment (jrs65) [reply setup.py]: numcodecs should work on macos now.

File: setup.py [hunk 37]
```diff
@@ -61,7 +61,15 @@
     """,
     python_requires=">=3.6",
     install_requires=requires,
-    extras_require={"mpi": ["mpi4py>=1.3"], "profiling": ["psutil", "pyinstrument"]},
+    extras_require={
+        "mpi": ["mpi4py>=1.3"],
+        "compression": [
+            "bitshuffle @ git+https://github.com/kiyo-masui/bitshuffle.git",
+            "numcodecs==0.7.3",
+            "zarr==2.8.1;platform_system!='Darwin'", # zarr is not supported on MacOS
```
Comment (jrs65) [inline setup.py:69]: Can remove conditional here as zarr should work on macOS again.

File: caput/tests/test_truncate.py (hunk 36 above)
[the remark this answers is not in the document]
Comment (jrs65) [reply caput/tests/test_truncate.py:1]: Let's add a few special cases in for this.

File: caput/memh5.py [hunk 38]
```diff
@@ -502,17 +577,69 @@ def to_hdf5(
...
+        """
         if not self.distributed:
-            with h5py.File(filename, mode, **kwargs) as f:
+            with file_format.open(filename, mode, **kwargs) as f:
                 deep_group_copy(
                     self,
                     f,
                     convert_attribute_strings=convert_attribute_strings,
                     convert_dataset_strings=convert_dataset_strings,
+                    file_format=file_format,
```
[the remark this answers is not in the document]
Comment (jrs65) [reply caput/memh5.py:638]: This seems to be very confusing. We should try to simplify.

File: caput/tests/test_selection_parallel.py (hunk 35 above)
Comment (jrs65) [reply caput/tests/test_selection_parallel.py:59]: A nicety, but not very important.

File: caput/mpiarray.py (hunk 33 above)
Comment (anjakefala) [reply caput/mpiarray.py:776]: Oh *wow*, they are not documented anywhere.

File: caput/config.py [hunk 39]
```diff
@@ -510,7 +510,7 @@ class Project:
 
             mode = file_format(default='zarr')
     """
-    options = ("hdf5", "zarr", None)
```
Comment (tristpinsm) [inline caput/config.py]: without a `None` default option, the `caput.fileformats.check_file_format` guessing functionality is broken.

File: caput/memh5.py (hunk 38 above)
[the remark this answers is not in the document]
Comment (jrs65) [reply caput/memh5.py:627]: Problem is with file_format, compression opts have been removed.

File: caput/mpiarray.py (hunk 33 above)
Comment (jrs65) [reply caput/mpiarray.py:776]: Still need doc entry for `chunks`

File: caput/tests/test_selection_parallel.py (hunk 35 above)
Comment (jrs65) [reply caput/tests/test_selection_parallel.py:59]: I can do this when I fix the tests.

File: caput/config.py (hunk 39 above)
Comment (jrs65) [reply caput/config.py]: @tristpinsm is looking into this.

**Review by jrs65 — APPROVED** [review]

**Comment by anjakefala** [comment]
:clap: :clap: :clap: :clap:

**Comment by jrs65** [comment]
> 👏 👏 👏 👏

Slowly getting there. Now I'm fighting to merge mpiarray in!!
